src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+```
+
+{:.bad}
+```dart
+@JS()
+external set list(List _);
+```
+
+These same requirements exist when you use [`Function.toJS`] to make a Dart
+function callable in JS. The values that flow into and out of this callback must
+be a compatible interop type or a primitive.
+
+If you use a Dart primitive e.g. `String`, an implicit conversion happens in the
+compiler to convert that value from a JS value to a Dart value. If performance
+is critical and you don’t need to examine the contents of the string, then using
+`JSString` instead to avoid the conversion cost may make sense.
+
+## Compatibility, type checks, and casts
+
+As mentioned above, the representation type of JS types may differ based on the
+compiler. This affects runtime type-checking and casts. Therefore, almost always
+avoid `is` checks where the value is an interop type or where the target type is
+an interop type. Avoid casts between Dart types and interop types. In order to
+type-check a JS value, use an interop member like [`typeofEquals`] or
+[`instanceOfString`] that examines the JS value itself.
+
+Bad:
+
+{:.bad}
+```dart
+void f(JSAny a) {
+  if (a is String) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSObject o) {
+  if (o is JSObject) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSString s) {
+  s as String;
+}
+```
+
+Okay:
+
+{:.good}
+```dart
+void f(JSAny a) {
+  // Here `a` is verified to be a JS function, so the cast is okay.
+  if (a.typeofEquals('function')) {
+    a as JSFunction;
+  }
+}
+```
+
+Dart may add lints to make runtime checks with JS interop types easier to avoid.
+
+## `null` vs `undefined`
+
+JS has both a `null` and an `undefined` value. This is in contrast with Dart,
+which only has `null`. In order to make JS values more ergonomic to use, if an
+interop member were to return either JS `null` or `undefined`, the compiler maps
+these values to Dart `null`. Therefore a member like:
+
+```dart
+@JS()
+external JSObject? get value;
+```
+
+can be interpreted as returning a JS object, JS `null`, or `undefined`.
+
+**Important**: There is a subtle inconsistency with regards to `undefined`
+between compiling to JS and Wasm. While compiling to JS *treats* `undefined`
+values as if they were Dart `null`, it doesn’t actually *change* the value
+itself. If an interop member returns `undefined` and you pass that value back
+into JS, JS will see `undefined` and not `null` when compiling to JS.
+However, when compiling to Wasm, this is not the case and the value will be
+`null` in JS. This is because the compiler *converts* the value to Dart `null`
+when compiling to Wasm, thereby losing information on whether the original value
+was JS `null` or `undefined`. Avoid writing code where this distinction matters
+by explicitly passing Dart `null` instead to an interop member. Currently,
+there's no platform-consistent way to provide `undefined` to interop members or
+distinguish between JS `null` and `undefined` values, but this will likely
+change in the future.
```

Comment:
Didn't rewrite anything, just added the site alert and broke up the paragraphs:

```suggestion
{{site.alert.warn}}
There is a subtle inconsistency with regards to `undefined`
between compiling to JS and Wasm. While compiling to JS *treats* `undefined`
values as if they were Dart `null`, it doesn’t actually *change* the value
itself. If an interop member returns `undefined` and you pass that value back
into JS, JS will see `undefined`, *not* `null`, when compiling to JS.

However, when compiling to Wasm, this is not the case and the value will be
`null` in JS. This is because the compiler *converts* the value to Dart `null`
when compiling to Wasm, thereby losing information on whether the original value
was JS `null` or `undefined`. Avoid writing code where this distinction matters
by explicitly passing Dart `null` instead to an interop member. 

Currently, there's no platform-consistent way to provide `undefined` to
interop members or distinguish between JS `null` and `undefined` values,
but this will likely change in the future.
{{site.alert.end}}
```

The part where it says "The compiler converts the value to Dart `null`...", would that be an _implicit conversion_?
Would it add anything to add "implicit" there? Just wondering since you talk about implicit conversions earlier it would be nice to make that call back, and since you mention explicit later here. 

If there's an issue where the "likely change in the future" is being tracked, link that here too!

Reply:
> would that be an implicit conversion? Would it add anything to add "implicit" there?

Indeed! This is a good example of an implicit conversion versus the explicit conversions users can do themselves.

Done wrt adding the issue.